app/src/main/java/org/luxoft/sdl_core/ClassicBtHandler.java

```diff
@@ -0,0 +1,375 @@
+package org.luxoft.sdl_core;
+
+import android.bluetooth.BluetoothAdapter;
+import android.bluetooth.BluetoothDevice;
+import android.bluetooth.BluetoothServerSocket;
+import android.bluetooth.BluetoothSocket;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.util.Log;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.UUID;
+
+public class ClassicBtHandler {
+    private final Context context;
+    private static ClassicBtHandler instance = null;
+    private BluetoothAdapter mBtAdapter = BluetoothAdapter.getDefaultAdapter();
+    private ConnectThread mConnectThread;
+    private ConnectedThread mConnectedThread;
+    private AcceptThread mAcceptThread;
+    private int mState;
+
+    // Name for the SDP record when creating server socket
+    private static final String NAME_INSECURE = "SDLCoreInsecure";
+
+    // Unique UUID for this application
+    private static final UUID SDL_UUID =
+            UUID.fromString("936DA01F-9ABD-4D9D-80C7-02AF85C822A8");
+
+    // Constants that indicate the current connection state
+    public static final int STATE_NONE = 0;       // we're doing nothing
+    public static final int STATE_LISTEN = 1;     // now listening for incoming connections
+    public static final int STATE_CONNECTING = 2; // now initiating an outgoing connection
+    public static final int STATE_CONNECTED = 3;  // now connected to a remote device
+
+    public static final String TAG = ClassicBtHandler.class.getSimpleName();
+
+    public static synchronized ClassicBtHandler getInstance(Context context) {
+        if (instance == null) {
+            instance = new ClassicBtHandler(context);
+        }
+        return instance;
+    }
+
+    private ClassicBtHandler(Context context) {
+        this.context = context;
+        // Register for broadcasts when a device is discovered.
+        IntentFilter filter = new IntentFilter(BluetoothDevice.ACTION_FOUND);
+        context.registerReceiver(mReceiver, filter);
+        mState = STATE_NONE;
+    }
+
+    public void DoDiscovery() {
+        Log.i(TAG, "If we're already discovering, stop it");
+        //
+        if (mBtAdapter.isDiscovering()) {
+            mBtAdapter.cancelDiscovery();
+        }
+        Log.i(TAG, "Request discover from BluetoothAdapter");
+        // Request discover from BluetoothAdapter
+        mBtAdapter.startDiscovery();
+    }
+
+    public void disconnect() {
+        Log.d(TAG, "Closing bluetooth handler...");
+        context.unregisterReceiver(mReceiver);
```

Comment:
@ychernysheva it is a good practice to make broadcast receiver registration/unregistration in opposite methods. In your case, if you want it to be unregistered on disconnect, then probably it should be registered on connect. Or, if you want registration in ctor, then you should have some `onDestroy` method